// @google-cloud/nodejs-repo-tools/cjs/cli/commands/exec.js
/**
 * Copyright 2017, Google, Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

require('colors');

var childProcess = require('child_process');

var buildPack = require('../../build_packs').getBuildPack();
var utils = require('../../utils');

var CLI_CMD = 'exec';
var COMMAND = `tools exec ${'[options] [--] <cmd> [args...]'.yellow}`;
var DESCRIPTION = `Run a given command in ${buildPack._cwd.yellow}.`;
var USAGE = `Usage:
  ${COMMAND.bold}
Description:
  ${DESCRIPTION}`;

exports.command = CLI_CMD;
exports.description = DESCRIPTION;
exports.builder = function (yargs) {
  yargs.usage(USAGE).example('Run the given command in the specified directory:').example(`- ${'tools exec -l=~/projects/some/dir -- echo "hi"'.cyan}`);
};

exports.handler = function (opts) {
  if (opts.dryRun) {
    utils.logger.log(CLI_CMD, 'Beginning dry run.'.cyan);
  }

  buildPack.expandConfig(opts);

  opts.args || (opts.args = []);

  if (opts.args.length === 0) {
    utils.logger.fatal(CLI_CMD, 'You must provide at least 1 command argument!');
  }

  utils.logger.log(CLI_CMD, `Executing in: ${opts.localPath.yellow}`);
  utils.logger.log(CLI_CMD, 'Running:', opts.args.join(' ').yellow);

  if (opts.dryRun) {
    utils.logger.log(CLI_CMD, 'Dry run complete.'.cyan);
    return;
  }

  var options = {
    cwd: opts.localPath,
    stdio: opts.silent ? 'ignore' : 'inherit',
    shell: true
  };

  var start = Date.now();

  childProcess.spawn(opts.args[0], opts.args.slice(1), options).on('exit', function (code, signal) {
    var timeTakenStr = utils.getTimeTaken(start);
    if (code !== 0 || signal) {
      utils.logger.error(CLI_CMD, `Oh no! Execution failed after ${timeTakenStr}.`);
      // eslint-disable-next-line no-process-exit
      process.exit(code || 1);
    } else {
      utils.logger.log(CLI_CMD, `Success! Execution finished in ${timeTakenStr}.`.green);
    }
  });
};
